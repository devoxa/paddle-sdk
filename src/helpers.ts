import { serialize as phpSerialize } from 'php-serialize'
import dayjs from 'dayjs'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import utc from 'dayjs/plugin/utc'
import FormData from 'form-data'

dayjs.extend(customParseFormat)
dayjs.extend(utc)

export function stableSerialize(object: { [key: string]: any }): string {
  // 1) Sort the object alphabetically by it's keys
  object = sortByKey(object)

  // 2) Encode arrays in their string form: `[1, 2, 3]` -> `'1, 2, 3'`
  // 3) Encode any non-strings as their JSON stringified version: `3` -> `'3'`
  for (const property in object) {
    if (object.hasOwnProperty(property) && typeof object[property] !== 'string') {
      if (Array.isArray(object[property])) {
        object[property] = object[property].join(', ')
      } else {
        object[property] = JSON.stringify(object[property])
      }
    }
  }

  // 4) Serialize in the way that PHP would
  return phpSerialize(object)
}

function sortByKey(object: { [key: string]: any }) {
  const keys = Object.keys(object).sort()

  const sortedObject: { [key: string]: any } = {}
  for (const i in keys) {
    sortedObject[keys[i]] = object[keys[i]]
  }

  return sortedObject
}

export function parseUtcDate(dateString: string, type: 'DATE' | 'DATE_TIME' | 'EXPIRY_DATE'): Date {
  switch (type) {
    case 'DATE':
      return dayjs.utc(dateString, 'YYYY-MM-DD').toDate()
    case 'DATE_TIME':
      return dayjs.utc(dateString, 'YYYY-MM-DD HH:mm:ss').toDate()
    case 'EXPIRY_DATE':
      return dayjs.utc(dateString, 'MM/YYYY').toDate()
  }
}

export function objectToFormData(object: { [key: string]: any }): FormData {
  const formData = new FormData()

  Object.entries(object).map(([key, value]) => {
    formData.append(key, value.toString())
  })

  return formData
}
