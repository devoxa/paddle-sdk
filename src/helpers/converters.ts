import dayjs from 'dayjs'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import utc from 'dayjs/plugin/utc'
import {
  PaddleSdkCardBrand,
  PaddleSdkCountry,
  PaddleSdkCurrency,
  PaddleSdkPausedReason,
  PaddleSdkPaymentMethod,
  PaddleSdkRefundType,
  PaddleSdkSubscriptionStatus,
} from '../interfaces'
import {
  RawPaddlePaymentRefundedAlert,
  RawPaddleSubscriptionCreatedAlert,
  RawPaddleSubscriptionPaymentSucceededAlert,
  RawPaddleSubscriptionUpdatedAlert,
} from '../__generated__/webhook-alerts'

dayjs.extend(customParseFormat)
dayjs.extend(utc)

export function convertApiInteger(integerString: string): number {
  return parseInt(integerString, 10)
}

export function convertApiFloat(floatString: string): number {
  return parseFloat(floatString)
}

export function convertApiBoolean(booleanString: '0' | '1' | 'false' | 'true'): boolean {
  return booleanString === '1' || booleanString === 'true'
}

export function convertSdkBoolean(boolean: boolean): 0 | 1 {
  return boolean ? 1 : 0
}

export function convertApiDate(
  dateString: string,
  type: 'DATE' | 'DATE_TIME' | 'EXPIRY_DATE'
): Date {
  switch (type) {
    case 'DATE':
      return dayjs.utc(dateString, 'YYYY-MM-DD').toDate()
    case 'DATE_TIME':
      return dayjs.utc(dateString, 'YYYY-MM-DD HH:mm:ss').toDate()
    case 'EXPIRY_DATE':
      return dayjs.utc(dateString, 'MM/YYYY').toDate()
  }
}

export function convertSdkDate(date: Date, type: 'DATE' | 'DATE_TIME' | 'EXPIRY_DATE'): string {
  switch (type) {
    case 'DATE':
      return dayjs.utc(date).format('YYYY-MM-DD')
    case 'DATE_TIME':
      return dayjs.utc(date).format('YYYY-MM-DD HH:mm:ss')
    case 'EXPIRY_DATE':
      return dayjs.utc(date).format('MM/YYYY')
  }
}

export function convertApiSubscriptionStatus(
  subscriptionStatus: RawPaddleSubscriptionCreatedAlert['status']
): PaddleSdkSubscriptionStatus {
  switch (subscriptionStatus) {
    case 'active':
      return PaddleSdkSubscriptionStatus.ACTIVE
    case 'trialing':
      return PaddleSdkSubscriptionStatus.TRIALING
    case 'past_due':
      return PaddleSdkSubscriptionStatus.PAST_DUE
    case 'paused':
      return PaddleSdkSubscriptionStatus.PAUSED
    case 'deleted':
      return PaddleSdkSubscriptionStatus.CANCELLED
  }
}

export function convertSdkSubscriptionStatus(
  subscriptionStatus: PaddleSdkSubscriptionStatus
): RawPaddleSubscriptionCreatedAlert['status'] {
  switch (subscriptionStatus) {
    case PaddleSdkSubscriptionStatus.ACTIVE:
      return 'active'
    case PaddleSdkSubscriptionStatus.TRIALING:
      return 'trialing'
    case PaddleSdkSubscriptionStatus.PAST_DUE:
      return 'past_due'
    case PaddleSdkSubscriptionStatus.PAUSED:
      return 'paused'
    case PaddleSdkSubscriptionStatus.CANCELLED:
      return 'deleted'
  }
}

export function convertApiPausedReason(
  pausedReason: Exclude<RawPaddleSubscriptionUpdatedAlert['paused_reason'], undefined>
): PaddleSdkPausedReason {
  switch (pausedReason) {
    case 'delinquent':
      return PaddleSdkPausedReason.DELINQUENT
    case 'voluntary':
      return PaddleSdkPausedReason.VOLUNTARY
  }
}

export function convertApiCurrency(currency: string): PaddleSdkCurrency {
  // These are the currencies already returned by paddle, we just make them type-safe
  return currency as PaddleSdkCurrency
}

export function convertApiCountry(country: string): PaddleSdkCountry {
  // These are the countries already returned by paddle, we just make them type-safe
  return country as PaddleSdkCountry
}

export function convertApiPaymentMethod(
  paymentMethod: RawPaddleSubscriptionPaymentSucceededAlert['payment_method']
): PaddleSdkPaymentMethod {
  switch (paymentMethod) {
    case 'card':
      return PaddleSdkPaymentMethod.CARD
    case 'paypal':
      return PaddleSdkPaymentMethod.PAYPAL
    case 'apple-pay':
      return PaddleSdkPaymentMethod.APPLE_PAY
    case 'wire-transfer':
      return PaddleSdkPaymentMethod.WIRE_TRANSFER
    case 'free':
      return PaddleSdkPaymentMethod.FREE
  }
}

export function convertApiCardBrand(cardBrand: string): PaddleSdkCardBrand {
  switch (cardBrand) {
    case 'visa':
      return PaddleSdkCardBrand.VISA
    case 'american_express':
      return PaddleSdkCardBrand.AMERICAN_EXPRESS
    case 'discover':
      return PaddleSdkCardBrand.DISCOVER
    case 'jcb':
      return PaddleSdkCardBrand.JCB
    case 'elo':
      return PaddleSdkCardBrand.ELO
    case 'master':
    case 'mastercard':
      return PaddleSdkCardBrand.MASTERCARD
    case 'maestro':
      return PaddleSdkCardBrand.MAESTRO
    case 'diners_club':
      return PaddleSdkCardBrand.DINERS_CLUB
  }

  return PaddleSdkCardBrand.UNKNOWN
}

export function convertApiRefundType(
  refundType: RawPaddlePaymentRefundedAlert['refund_type']
): PaddleSdkRefundType {
  switch (refundType) {
    case 'full':
      return PaddleSdkRefundType.FULL
    case 'vat':
      return PaddleSdkRefundType.VAT
    case 'partial':
      return PaddleSdkRefundType.PARTIAL
  }
}

export function convertSdkPriceList(
  currencyList: Array<[PaddleSdkCurrency, number]>
): Array<string> {
  return currencyList.map(([currency, amount]) => `${currency}:${amount}`)
}
