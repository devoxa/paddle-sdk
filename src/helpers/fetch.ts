import nodeFetch from 'node-fetch'
import FormData from 'form-data'

type Body = Record<string, string | number | undefined>

export async function fetch<T>(url: string, options: { method: string; body: Body }): Promise<T> {
  const response = await nodeFetch(url, {
    method: options.method,
    body: objectToFormData(options.body),
  })

  return response.json()
}

function objectToFormData(object: Body): FormData {
  const formData = new FormData()

  Object.entries(object).forEach(([key, value]) => {
    if (typeof value === 'undefined') return

    formData.append(key, value.toString())
  })

  return formData
}
