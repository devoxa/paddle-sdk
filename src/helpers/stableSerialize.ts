import { serialize as phpSerialize } from 'php-serialize'

export function stableSerialize<T>(object: Record<string, T>): string {
  // 1) Sort the object alphabetically by it's keys
  object = sortByKey(object)

  // 2) Encode arrays in their string form: `[1, 2, 3]` -> `'1, 2, 3'`
  // 3) Encode any non-strings as their JSON stringified version: `3` -> `'3'`
  const encodedObject: Record<string, string> = {}
  for (const property in object) {
    if (object.hasOwnProperty(property) && typeof object[property] !== 'string') {
      const value = object[property]

      if (Array.isArray(value)) {
        encodedObject[property] = value.join(', ')
      } else {
        encodedObject[property] = JSON.stringify(value)
      }
    }
  }

  // 4) Serialize in the way that PHP would
  return phpSerialize(object)
}

function sortByKey<T>(object: Record<string, T>) {
  const keys = Object.keys(object).sort()

  const sortedObject: Record<string, T> = {}
  for (const i in keys) {
    sortedObject[keys[i]] = object[keys[i]]
  }

  return sortedObject
}
