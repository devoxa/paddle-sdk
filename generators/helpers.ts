import fetch from 'node-fetch'

type NextData = {
  props: {
    pageProps: {
      dehydratedState: {
        queries: Array<NextDehydratedStateQuery>
      }
    }
  }
}

type NextDehydratedStateQuery = {
  queryHash: string
  state: {
    data: {
      data: string
    }
  }
}

export async function fetchPageMarkdown(url: string) {
  const response = await fetch(url)
  const text = await response.text()

  const scriptMatch = text.match(/<script id="__NEXT_DATA__" [^>]*>(.*?)<\/script>/)
  if (!scriptMatch) throw new Error('Could not find __NEXT_DATA__ script')

  const data = JSON.parse(scriptMatch[1]) as NextData
  const queries = data.props.pageProps.dehydratedState.queries

  const urlQueryKey = url.split('/').pop() as string
  const queryHashRegex = new RegExp(`"[a-z0-9]{13}-${urlQueryKey}`)
  const query = queries.find((query) => query.queryHash.match(queryHashRegex))
  if (!query) throw new Error(`Could not find query for "${urlQueryKey}" in dehydratedState`)

  return query.state.data.data
}

export function spliceAndReturn<T>(array: Array<T>, index: number): T {
  const element = array[index]
  array.splice(index, 1)
  return element
}

export function trim(string: string, character: string) {
  let start = 0
  let end = string.length

  while (start < end && string[start] === character) ++start
  while (end > start && string[end - 1] === character) --end

  return start > 0 || end < string.length ? string.slice(start, end) : string
}

export function toPascalCase(string: string) {
  const camelCase = string.replace(/_(\w)/g, (_, $1) => $1.toUpperCase())
  return `${camelCase.charAt(0).toUpperCase()}${camelCase.slice(1)}`
}

export function toJSDoc(lines: Array<string>) {
  return lines.length ? `/**\n${lines.map((line) => `* ${line}`).join('\n')}\n*/` : ''
}
