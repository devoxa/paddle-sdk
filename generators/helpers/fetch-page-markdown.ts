import fetch from 'node-fetch'

type NextData = {
  props: {
    pageProps: {
      dehydratedState: {
        queries: Array<NextDehydratedStateQuery>
      }
    }
  }
}

type NextDehydratedStateQuery = {
  queryHash: string
  state: {
    data: {
      data: string
    }
  }
}

export async function fetchPageMarkdown(url: string) {
  const response = await fetch(url)
  const text = await response.text()

  const scriptMatch = text.match(/<script id="__NEXT_DATA__" [^>]*>(.*?)<\/script>/)
  if (!scriptMatch) throw new Error('Could not find __NEXT_DATA__ script')

  const data = JSON.parse(scriptMatch[1]) as NextData
  const queries = data.props.pageProps.dehydratedState.queries

  const urlQueryKey = url.split('/').pop() as string
  const query = queries.find((query) => query.queryHash.includes(urlQueryKey))
  if (!query) throw new Error(`Could not find query for "${urlQueryKey}" in dehydratedState`)

  return query.state.data.data
}
