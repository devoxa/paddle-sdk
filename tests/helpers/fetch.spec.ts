import { fetch } from '../../src/helpers/fetch'
import FormData from 'form-data'
import nodeFetch from 'node-fetch'

jest.mock('node-fetch', () => jest.fn())

function mockNextApiResponse(json: any) {
  ;((nodeFetch as unknown) as jest.Mock).mockImplementationOnce(async () => ({
    json: async () => json,
  }))
}

function getLastApiRequest() {
  return ((nodeFetch as unknown) as jest.Mock).mock.calls[0]
}

describe('fetch', () => {
  beforeEach(() => {
    ;((nodeFetch as unknown) as jest.Mock).mockClear()
  })

  it('runs fetch using form data', async () => {
    mockNextApiResponse({ success: true, response: { foo: 'bar' } })

    const response = await fetch('https://example.com', { method: 'POST', body: { baz: 42 } })
    expect(response).toMatchSnapshot()

    const request = getLastApiRequest()
    expect(request[0]).toEqual('https://example.com')
    expect(request[1].method).toEqual('POST')
    expect(request[1].body).toBeInstanceOf(FormData)
  })
})
