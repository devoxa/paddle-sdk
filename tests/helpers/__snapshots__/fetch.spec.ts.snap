// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch runs fetch using form data 1`] = `
Object {
  "response": Object {
    "foo": "bar",
  },
  "success": true,
}
`;
